New Tutorial: REST API

bilgeyucel opened the pull request:
In this tutorial, I tried to explain the whole process from setting up an environment to querying the pipeline using REST API.

This tutorial cannot be run in Colab, therefore, the structure is different.

I created a [preview link](https://haystack-home-qm0vo1ard-deepset-overnice.vercel.app/tutorials/20_using_haystack_with_rest_api) if you want to see how it is displayed on Haystack website.

TODO:

- [x] Update README - waiting for tutorial 19 to be merged
- [x] Remove Colab link - this tutorial cannot be run on Google Colab, removing the link is possible after #81 
- [x] Merge `main` and generate the .md file with the download button and new title convention

agnieszka-m commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,485 @@
+   "source": [
+    "## Create your pipeline\n",
+    "\n",
+    "Now you know how REST API works, you can start customizing the pipeline"
```
```suggestion
    "Now that you know how Haystack REST API works, you can start customizing the search system."
```

One note - please pay attention if you're talking to the user through "we" or "you". I think the rest of hte tutorials uses "you" (at least the restructured ones) so I'd be for this form. This means that instead of saying "we'll initiate a document store now" you'll say "NOw, you'll initiate a document store"

agnieszka-m commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,485 @@
+   "source": [
+    "### Start Elasticsearch \n",
+    "\n",
+    "Replace the Docker image with your image or use suitable Elasticsearch image. Use `image: \"docker.elastic.co/elasticsearch/elasticsearch:7.9.2\"` for an empty DocumentStore."
```
```suggestion
    "Replace the Docker image with your image or use an Elasticsearch image. To use an empty DocumentStore, run: `image: \"docker.elastic.co/elasticsearch/elasticsearch:7.9.2\"`e."
```

agnieszka-m commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,485 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "* Go to your `docker-compose.yml` file\n",
```
```suggestion
    "1.Go to your `docker-compose.yml` file.\n",
```

agnieszka-m commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,485 @@
+   "metadata": {},
+   "source": [
+    "* Go to your `docker-compose.yml` file\n",
+    "* Replace line 22 with `image: \"docker.elastic.co/elasticsearch/elasticsearch:7.9.2\"`\n",
```
```suggestion
    "2. Replace line 22 with `image: \"docker.elastic.co/elasticsearch/elasticsearch:7.9.2\".`\n",
```

agnieszka-m commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,485 @@
+   "source": [
+    "* Go to your `docker-compose.yml` file\n",
+    "* Replace line 22 with `image: \"docker.elastic.co/elasticsearch/elasticsearch:7.9.2\"`\n",
+    "* Restart Elasticsearch"
```
```suggestion
    "3. Restart Elasticsearch."
```
Whenever there's a sequence of steps, we use a numbered list. Unordered lists are only for listing items whose order doesn't matter. List items always end with a full stop. If you need more info, have a look at our writing guidelines :).

agnieszka-m commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,485 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The new `docker-compose.yml` file should look like this:\n",
```
```suggestion
    "Here's what your new `docker-compose.yml` file should look like:\n",
```

agnieszka-m commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,485 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### New document search pipeline\n",
```
```suggestion
    "### Create a Document Search Pipeline\n",
```

The tutorial is all about tasks so we want the headers to be task based. We use Chicago-style header capitalization, which means everythig except for prepositions is capitalized. You can always check at https://capitalizemytitle.com/style/Chicago/.

agnieszka-m commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,485 @@
+   "source": [
+    "### New document search pipeline\n",
+    "\n",
+    "We are going to create a document search pipeline. Create a new file named `document-search.haystack-pipeline.yml` in `/pipeline` folder. Do not forget to update `PIPELINE_YAML_PATH` in `config.py` with the new file name.\n",
```
```suggestion
    "Create a document search pipeline:
    1. Create a file named `document-search.haystack-pipeline.yml` in the `/pipeline` folder: (here comes the code).
    2. Update `PIPELINE_YAML_PATH` in `config.py` with the new file name.\n",
```
Here's why: it's a sequence, so ordered list. Our voice: time to value, positive language (hence: Update config.py instead of "Don't forget to..").

agnieszka-m commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,485 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "If you are using Docker Compose for Haystack API, there are further changes you need to make in `docker-compose.yml`:\n",
```
```suggestion
    "If you are using Docker Compose for Haystack API, update `docker-compose.yml`:\n",
```
Our voice: time to value (=be concise whenever possible)

agnieszka-m commented on tutorials/20_REST_API.ipynb:
````diff
@@ -0,0 +1,485 @@
+    "          - /<path_to_haystack>/haystack/rest_api/rest_api/pipeline:/home/user/rest_api/pipeline\n",
+    "    ...\n",
+    "    ``` \n",
+    "2. Update the `PIPELINE_YAML_PATH` value as `/home/user/rest_api/pipeline/document-search.haystack-pipeline.yml`"
````
```suggestion
    "2. Update the `PIPELINE_YAML_PATH` value to `/home/user/rest_api/pipeline/document-search.haystack-pipeline.yml`"
```

agnieszka-m commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,485 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This document search pipeline only requires a Retriever. Therefore, it is going to be enough for our query pipeline if we declare a DocumentStore and Retriever in `document-search.haystack-pipeline.yml`. \n",
```
```suggestion
    "This document search pipeline only needs a Retriever so all you need to declare in in `document-search.haystack-pipeline.yml` is the DocumentStore and the Retriever. \n",
```
Our voice: conversational, normal vs formal words

agnieszka-m commented on tutorials/20_REST_API.ipynb:
````diff
@@ -0,0 +1,485 @@
+    "        inputs: [Query]\n",
+    "...\n",
+    "```\n",
+    "Be sure to have the same `name` with `QUERY_PIPELINE_NAME` variable in `config.py`."
````
```suggestion
    "Make sure you update the `name` in the `QUERY_PIPELINE_NAME` variable in `config.py` to `query`."
```

agnieszka-m commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,485 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Indexing pipeline\n",
```
```suggestion
    "### Create an Indexing Pipeline\n",
```

brandenchan replied:
If we are going to suggest performing indexing via the /file-upload endpoint, I think we need to give a bit more guidance as to what this actually entails. What does the curl command look like?

agnieszka-m commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,485 @@
+   "source": [
+    "### Indexing pipeline\n",
+    "\n",
+    "You can use REST API to index your files to your document store. This requires an indexing pipeline. Add the indexing pipeline to `document-search.haystack-pipeline.yml`, then, you can use `/file-upload` endpoint to upload your files to Elasticsearch. \n",
```
```suggestion
    "You can use REST API to index your files to your document store. This requires an indexing pipeline. Add the indexing pipeline to `document-search.haystack-pipeline.yml` and use the `/file-upload` endpoint to upload your files to Elasticsearch. \n",
```

agnieszka-m commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,485 @@
+    "\n",
+    "You can use REST API to index your files to your document store. This requires an indexing pipeline. Add the indexing pipeline to `document-search.haystack-pipeline.yml`, then, you can use `/file-upload` endpoint to upload your files to Elasticsearch. \n",
+    "\n",
+    "Download the same demo files [here](https://s3.eu-central-1.amazonaws.com/deepset.ai-farm-qa/datasets/documents/article_txt_countries_and_capitals.zip) and upload them using cURL. Check [this](https://docs.haystack.deepset.ai/docs/rest_api#indexing-documents-in-the-haystack-rest-api-document-store) documentation page for more detail about file indexing.\n",
```
```suggestion
    "Download the same [demo files](https://s3.eu-central-1.amazonaws.com/deepset.ai-farm-qa/datasets/documents/article_txt_countries_and_capitals.zip) you used in the first part of this tutorial and upload them using cURL. 
To learn more about indexing, see [Indexing Documents](https://docs.haystack.deepset.ai/docs/rest_api#indexing-documents-in-the-haystack-rest-api-document-store).\n",
```

agnieszka-m commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,485 @@
+    "Download the same demo files [here](https://s3.eu-central-1.amazonaws.com/deepset.ai-farm-qa/datasets/documents/article_txt_countries_and_capitals.zip) and upload them using cURL. Check [this](https://docs.haystack.deepset.ai/docs/rest_api#indexing-documents-in-the-haystack-rest-api-document-store) documentation page for more detail about file indexing.\n",
+    "\n",
+    "<aside>\n",
+    "⚠️ If you want to index your files directly to Elasticsearch through script, be sure to provide the same indexing pipeline with your `document-search.haystack-pipeline.yml` file for consistency between indexed files.\n",
```
```suggestion
    "⚠️To index your files directly to Elasticsearch through a script, use the same indexing pipeline with your `document-search.haystack-pipeline.yml` file to make sure the files indexed are consistent.\n",
```

Our voice: conversational

brandenchan replied:
I think I get what this is saying, but it's a bit hard to understand. Do you mean that if someone is doing both index via REST API endpoint and index via script that calls indexing pipeline, that the pipeline in each case is the same so that the indexed files are consistent? 

On a formatting level, I think this "aside" block is quite hard to distinguish from regular text on the rendered page and would opt against using it.

agnieszka-m commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,485 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The indexing pipeline should be as follows:\n",
```
```suggestion
    "Here's what the indexing pipeline should look like:\n",
```

agnieszka-m commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,485 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "When you restart the Haystack API, REST API is going to load the new pipeline and the pipeline will be ready to use. "
```
```suggestion
    "When you restart the Haystack API, REST API will load the new pipeline and the pipeline will be ready to use. "
```

agnieszka-m commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,485 @@
+ },
+ "nbformat": 4,
+ "nbformat_minor": 2
+}
```
I'd add a paragraph summing up what they have learned/built/achieved in this tutorial. Maybe some links they may want to check, additional info, etc.

agnieszka-m reviewed (changes requested):
Here are my 2 cents. I tried to add a reason for most of my comments,  but if anyting is unclear, just ping me and I'll be happy to chat. 
One thing that's not clear to me is why did we need to first set up the demo files and pipeline if we created a totally new pipeline after that?
Also, I think I'd maybe add more info about the pipeline that they were creating. For example, what params it's using and why (just one sentence like: this pipeline uses the default params that work well for htis amount of files blahblah)?

brandenchan commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,485 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Start Elasticsearch \n",
```
To be honest I found this section a bit confusing and I think it could be clearer. Why do they have to change the ES image? 

Also I think they could do with more explanation as to how to restart Elasticsearch - let's give them a specific command to run.

I also think telling them to change a specific line by number is dangerous - if a line is added or removed, this number is no longer valid.

brandenchan commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,485 @@
+   "source": [
+    "# Haystack with REST API\n",
+    "\n",
+    "- **Level**: Intermediate\n",
```
To be honest, I think this might actually be an advanced topic. The user has to work with Docker, requests to REST API, pipeline ymls. They also have to make changes to lots of files which they probably don't fully understand the significance of.

brandenchan commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,485 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Voilà! Make a new query!\n",
```
Can we show users what the expected response should look like?

brandenchan reviewed (changes requested):
Thanks for making a start on this! I think that you have covered many of the start up steps for the REST API in good detail. 

I've made a set of comments that mostly request structural / content changes that I think will simplify the tutorial or provide users with a little bit of conceptual overview. This tutorial covers a really tough topic because we are asking users to interact a lot of components (e.g. Docker, CURL, Yaml Pipelines, REST API, Swagger) and I would push to make the experience still a bit more straight forward for the user. 

More than happy to clarify any thing that's unclear, or to have a chat to align!

TuanaCelik commented on markdowns/20_REST_API.md:
```diff
@@ -0,0 +1,319 @@
+- **Time to complete**: 30 minutes
+- **Prerequisites**: N/A
+- **Nodes Used**: `ElasticsearchDocumentStore`, `EmbeddingRetriever`
+- **Goal**: After completing this tutorial, you will have learned how you can interact with Haystack through REST API.
```
'interact with a Haystack pipeline' ?

TuanaCelik commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,485 @@
+    "\n",
+    "- **Level**: Intermediate\n",
+    "- **Time to complete**: 30 minutes\n",
+    "- **Prerequisites**: N/A\n",
```
Honestly I would suggest making this tutorial condensed to 'using the Rest API'. Lmk what you think here as I think this might be a bit controversial:
The usual problems we see is simply technical, in terms of just setting up, using and interacting with docker and rest api. So I would argue understanding in detail the pipeline structure and each component isn't the aim here. Imo the things people should come away with is:
- These .yml files are where I define my config
- They are able to edit the file in a way, maybe with some examples (change the model, reader setting etc) and use this as the pipeline that they query with rest api
- For details on each NLP related component and par I would forward them to the relevant docs (e.g. the full set of options for a preprocessor, or reader, or retriever etc)

Adding this comment as a reply here because of the mention of 'NLP concepts' which I think _could_ be added as a prerequisite. But mainly I would say the prerequisites are: basic understanding of the YAML syntax (link to resource), basic understanding of docker.

bilgeyucel commented:
Hi @TuanaCelik, @brandenchan, and @agnieszka-m, I restructured the tutorial according to your feedback. The main changes I made:
* Removed the first part where I explained the demo setup. Now I directly focus on creating a new application but with more guidance 
* Didn't mention Swagger UI at all. Swagger UI is nice to use but not a production-focused tool, so, I removed it completely
* Added more explanation to all steps. 
* Followed the new structure of tutorials on #44 

I think this version is cleaner and easier to follow. I tried to pay attention not to suggest options and tried to stick with one approach only. 

Let me know what you think! 🙌

anakin87 commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,538 @@
+    "pip install 'farm-haystack[all]'\n",
+    "pip install -e rest_api/\n",
+    "\n",
+    "brew install xpdf # required for PDFToTextConverter node"
```
Sorry for the intrusion... by the way, great job!

I think that `brew` is installed by default only in macOS, not in Linux and Windows.

From [Haystack workflows for tests](https://github.com/deepset-ai/haystack/blob/0e05f71f330ab0ad14541efa64f2ba95ce87c2e6/.github/workflows/tests.yml#L461), I see that you can install `xpdf` on Linux with the following command:
`wget --no-check-certificate https://dl.xpdfreader.com/xpdf-tools-linux-4.04.tar.gz && tar -xvf xpdf-tools-linux-4.04.tar.gz && sudo cp xpdf-tools-linux-4.04/bin64/pdftotext /usr/local/bin`

For Windows you can install `xpdf` with `choco install xpdf-utils`, but you need to have Chocolatey.

brandenchan commented on markdowns/20_REST_API.md:
```diff
@@ -0,0 +1,370 @@
+
+2. Index Files to Elasticsearch
+
+You can send cURL requests to `/file-upload` endpoint to upload files to Elasticsearch instance. Replace `<PATH_TO_FOLDER>` with the path to the example files on your computer and send the POST request. If the file is successfully uploaded, you will get a response as `null`.
```
```suggestion
You can send cURL requests to the `/file-upload` endpoint to upload files to the Elasticsearch instance. Replace `<PATH_TO_FOLDER>` with the path to the example files on your computer and send the POST request. If the file is successfully uploaded, you will get a response as `null`.
```

brandenchan commented on markdowns/20_REST_API.md:
````diff
@@ -0,0 +1,370 @@
+     --form meta=null
+```
+
+However, this method is not convenient to upload multiple files to Elasticsearch instance as replacing the file names in the request by hand is difficult. Instead, create a python file in the folder that you keep the example files, put the code below into the python file, and run the python script. This python code takes the name of every file in the folder and sends the POST request to the url ` ...
````
```suggestion
However, this method is not convenient for uploading multiple files to the Elasticsearch instance as replacing the file names in the request by hand is difficult. Instead, create a python file in the folder that you keep the example files, put the code below into the python file, and run the python script. This python code takes the name of every file in the folder and sends the POST request to the url `http://127.0.0.1:8000/file-upload`. With the `print(response.text, ":", file_name)` line, you will be able to see the names of all indexed files. Make sure you see the "Completed" text before continuing. 
```

brandenchan commented on markdowns/20_REST_API.md:
````diff
@@ -0,0 +1,370 @@
+     }'
+```
+
+ As response, you will get a `QueryResponse` object consists of `query`, `answers`, and `documents`. Related [Documents](https://docs.haystack.deepset.ai/docs/documents_answers_labels#document) to your query will be under `documents` attribute of the object.
````
```suggestion
 As response, you will get a `QueryResponse` object consists of `query`, `answers`, and `documents`. Related [Documents](https://docs.haystack.deepset.ai/docs/documents_answers_labels#document) to your query will be under the `documents` attribute of the object.
```

brandenchan reviewed (changes requested):
The structure of this Tutorial is much much clearer! Good work. I have just made a set of language changes.

TuanaCelik commented on markdowns/20_REST_API.md:
```diff
@@ -0,0 +1,370 @@
+
+## Overview
+
+Haystack enables you to apply the latest NLP technology to your own data and create production-ready applications. Building an end-to-end NLP application requires the combination of multiple concepts. Here are those consepts:
```
Agreed!

TuanaCelik commented on markdowns/20_REST_API.md:
```diff
@@ -0,0 +1,370 @@
+
+Haystack enables you to apply the latest NLP technology to your own data and create production-ready applications. Building an end-to-end NLP application requires the combination of multiple concepts. Here are those consepts:
+* **DocumentStore** stores the data. You will use Elasticsearch for this tutorial.
+* **Haystack** pipelines convert documents, index them to DocumentStore, and run NLP tasks such as question answering and document search.
```
convert files in to **Documents** or Haystack **Documents** --> just because this might be a good opportunity for an aha moment that the learner gets about what their files are processed into

TuanaCelik commented on markdowns/20_REST_API.md:
```diff
@@ -0,0 +1,370 @@
+
+## Create Pipeline YAML File
+
+YAML files are widely used for confugurations. Haystack enables defining pipelines as YAML files and `load_from_yaml` method loads pipelines from YAML file. In a YAML file, `components` section defines all pipeline nodes and `pipelines` section defines how these nodes are connected to each other to form a pipeline. Let's start with defining query and indexing pipelines.
```
`load_from_yaml` lets you load the YAML pipeline into your python script or pipeline object right? If I have this right could the following make sense?

YAML files are widely used for configurations. Haystack enables defining pipelines as YAML files and the `load_from_yaml` method will even allow you to load a pipeline defined in YAML into a Python object.

bilgeyucel replied:
Hi @TuanaCelik, it is correct that it loads the YAML pipeline into a python object but it is done behind the scenes, so it doesn't explicitly allow them to load the pipeline in this tutorial, it is handled for them already. Here, I wanted to explain a bit about what happens in Haystack when a YAML pipeline is used. What do you think about these two suggestions? 

```suggestion
YAML files are widely used for configurations. Haystack enables defining pipelines as YAML files by using the `load_from_yaml` method that loads pipelines defined in YAML into Python objects. In a YAML file, the `components` section defines all pipeline nodes and the `pipelines` section defines how these nodes are connected to each other to form a pipeline. Let's start with defining query and indexing pipelines.
```

```suggestion
YAML files are widely used for configurations. Haystack enables defining pipelines as YAML files by using the `load_from_yaml` method. The `load_from_yaml` method loads pipelines defined in YAML into Python objects. In a YAML file, the `components` section defines all pipeline nodes and the `pipelines` section defines how these nodes are connected to each other to form a pipeline. Let's start with defining query and indexing pipelines.
```

TuanaCelik commented on markdowns/20_REST_API.md:
```diff
@@ -0,0 +1,370 @@
+
+1. Create a Document Search Pipeline
+
+As the query pipeline, you will create a Document Search pipeline from scratch. Create a new file named `document-search.haystack-pipeline.yml` in `/pipeline` folder under `/rest_api` in Haystack code base. Then, update `PIPELINE_YAML_PATH` value in `rest_api/config.py` with the new file name. `PIPELINE_YAML_PATH` will tell the REST API which YAML file to run. 
```
The 'As your query pipeline' seems off to me. How about:
"To act as your query pipeline, you will create.... "

TuanaCelik commented on markdowns/20_REST_API.md:
```diff
@@ -0,0 +1,370 @@
+    type: EmbeddingRetriever
+    params:
+      document_store: DocumentStore
+      top_k: 5 
```
This is up to you, but I think the section just under these yaml definitions are a great opportunity to hint to the learner how they can change setting. It might serve as a good mental map from the API reference they can see in docs to implementing them in a YAML file. `top_k` for example but you can in a short paragraph maybe hint them into some other things they could set?

I'll leave this decision up to you and @brandenchan depending on the level of 'optionality' you decide to provide in this tutorial

bilgeyucel replied:
Good idea to mention API reference 💡  I don't want to offer any options such as saying that "add `max_seq_len` to Retriever", instead, encourage them to play with the pipeline later and refer to the API Reference for that. So, this sentence👇 might come after sharing the whole YAML file. WDYT? @TuanaCelik @brandenchan  

> Feel free to play with the pipeline setup later on. Add or remove some nodes, change the parameters or add new ones. Check out [Haystack API Reference](https://docs.haystack.deepset.ai/reference/answer-generator-api) for more options on nodes and parameters.

TuanaCelik reviewed (commented):
Amazing job @bilgeyucel such great effort <3 
I have left some minor comments. And one extra one that you could consider with the help of someone in core-engineering maybe is to include what OS these commands would run on, or some type of pre-req (in terms of setup, additionally to the ones you've already provided)

bilgeyucel commented on markdowns/20_REST_API.md:
```diff
@@ -0,0 +1,370 @@
+## Overview
+
+Haystack enables you to apply the latest NLP technology to your own data and create production-ready applications. Building an end-to-end NLP application requires the combination of multiple concepts. Here are those consepts:
+* **DocumentStore** stores the data. You will use Elasticsearch for this tutorial.
```
Here it confuses me to put it like that. I said "You will use Elasticsearch for this tutorial" because they will start an Elasticsearch instance and then, use ElasticsearchDocumentStore to connect to the instance. Would it be better like this? 👇  
```suggestion
* **DocumentStore** stores the data. You will start an Elasticsearch instance and use `ElasticsearchDocumentStore` to connect to the instance.
```

bilgeyucel commented on markdowns/20_REST_API.md:
````diff
@@ -0,0 +1,370 @@
+        inputs: [Retriever]
+```
+
+## Start Haystack API
````
Tbh, I explained how to run it with Docker in the first version as an alternative to gunicorn so that people know it's possible. However, I couldn't get a stable development environment with Docker (it couldn't connect to elasticsearch, gave errors etc.), which prevented me to interact easily with the pipeline YAML file. Since this tutorial is about how to interact with REST API, I wanted to focus on that part instead of explaining how to run it with the current `docker-compose.yml` file as there are some additional steps there.

bilgeyucel commented on markdowns/20_REST_API.md:
````diff
@@ -0,0 +1,370 @@
+        inputs: [Retriever]
+```
+
+After completing query and indexing pipelines, add `version: ignore` to the top of the file and the pipeline YAML file is ready to run by Haystack API.
````
This suggested sentence states that Haystack API is ready but actually it is not. If we'd like to remove passive voice there, let's just say that it's completed, and be done. WDYT?

```suggestion
After completing query and indexing pipelines, add `version: ignore` to the top of the file. Now, the pipeline YAML is completed.
```

agnieszka-m commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,538 @@
+    "* **DocumentStore** stores the data. You will use Elasticsearch for this tutorial.\n",
+    "* **Haystack** pipelines convert files into [Documents](https://docs.haystack.deepset.ai/docs/documents_answers_labels#document), index them to DocumentStore, and run NLP tasks such as question answering and document search.\n",
+    "* **REST API** allows other applications to interact with Haystack by handling their queries and returning responses.\n",
+    "* **Docker** simplifies the environment set-up needed to have Elasticsearch running.\n",
```
```suggestion
    "* **Docker** simplifies the environment setup needed to have Elasticsearch running.\n",
```

agnieszka-m commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,538 @@
+    "* **REST API** allows other applications to interact with Haystack by handling their queries and returning responses.\n",
+    "* **Docker** simplifies the environment set-up needed to have Elasticsearch running.\n",
+    "\n",
+    "This tutorial introduces you to all the concepts needed to build an end-to-end document search application. After completing this tutorial, you will have learned how to create a pipeline YAML file, index files and query your application using REST API."
```
```suggestion
    "This tutorial introduces you to all the concepts needed to build an end-to-end document search application. After completing this tutorial, you will have learned how to create a pipeline YAML file, index files, and how to query your application using REST API."
```

agnieszka-m commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,538 @@
+   "source": [
+    "5. Launch Elasticsearch.\n",
+    "\n",
+    "Go to the directory where `docker-compose.yml` is and start the Elasticsearch container. You might have realised that there are settings for haystack and ui containers provided in the `docker-compose.yml`. They are not necessary for in this tutorial. Run `docker-compose up` with the `elasticsearch` argument so that only the Elasticsearch container launches."
```
```suggestion
    "Go to the directory where `docker-compose.yml` is and start the Elasticsearch container. Run `docker-compose up` with the `elasticsearch` argument so that only the Elasticsearch container launches."
```

They don't need these settings now so I'd say it's better not to mention them at al otherwise, it's confusing.

agnieszka-m commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,538 @@
+   "source": [
+    "## Create Pipeline YAML File\n",
+    "\n",
+    "YAML files are widely used for configurations. Haystack enables defining pipelines as YAML files and the `load_from_yaml()` method loads pipelines from YAML file. In a YAML file, the `components` section defines all pipeline nodes and `pipelines` section defines how these nodes are connected to each other to form a pipeline. Let's start with defining query and indexing pipelines."
```
```suggestion
    "YAML files are widely used for configurations. You can define Haystack pipelines as YAML files and then use the `load_from_yaml()` method to load pipelines from the YAML file. In a YAML file, the `components` section defines all pipeline nodes and the `pipelines` section defines how these nodes are connected to each other to form a pipeline. Let's start with defining query and indexing pipelines."
```

It's recommended not to say your product "enables" people to do something. Instead, you can say they can do something with the product.

bilgeyucel replied:
Hi @agnieszka-m, here, the user doesn't use the `load_from_yaml()` method, it is used by Haystack. I wanted to explain the process a little bit so that they can see the connection between a YAML file and python objects we use in Haystack. What do you think about this 👇 ?

```suggested
YAML files are widely used for configurations. You can define Haystack pipelines as YAML files and then `load_from_yaml()` method loads the pipeline defined in YAML into a Python object. In a YAML file, the `components` section defines all pipeline nodes and the `pipelines` section defines how these nodes are connected to each other to form a pipeline. Let's start with defining query and indexing pipelines.
```

agnieszka-m commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,538 @@
+   "source": [
+    "1. Create a document search pipeline.\n",
+    "\n",
+    "To act as your query pipeline, you will design a document search pipeline from scratch. Create a new file named `document-search.haystack-pipeline.yml` in `/pipeline` folder under `/rest_api` in the Haystack code base. Then, update the `PIPELINE_YAML_PATH` value in `rest_api/config.py` with the new file name. `PIPELINE_YAML_PATH` variable will tell the REST API which YAML file to run. \n",
```
```suggestion
    "Time to design a document search pipeline from scratch. This will be your query pipeline. Create a new file named `document-search.haystack-pipeline.yml` in the `/pipeline` folder under `/rest_api` in the Haystack code base. Then, update the `PIPELINE_YAML_PATH` value in `rest_api/config.py` with the new file name. `PIPELINE_YAML_PATH` variable will tell the REST API which YAML file to run. \n",
```

agnieszka-m commented on tutorials/20_REST_API.ipynb:
````diff
@@ -0,0 +1,538 @@
+    ")\n",
+    "```\n",
+    "\n",
+    "A document search pipeline requires a DocumentStore and a Retriever. Use `ElasticsearchDocumentStore` and `EmbeddingRetriever` for these nodes respectively and define them under `components`. For each component, `type` needs to refer to a node class in Haystack, `name` refers how these nodes are called in the pipeline YAML, and `params` is used for the parameters that can be provided to the  ...
````
```suggestion
    "A document search pipeline requires a DocumentStore and a Retriever. Use `ElasticsearchDocumentStore` and `EmbeddingRetriever` for these nodes respectively and define them under `components`. For each component, set `type` to a node class in Haystack, set `name` to how you want to call this node in the pipeline YAML, and `params` to set the node parameters. As Retriever parameters, provide `document_store`, `embedding_model`, and a `top_k` value. Let's start by defining the pipeline nodes:\n",
```

agnieszka-m commented on tutorials/20_REST_API.ipynb:
````diff
@@ -0,0 +1,538 @@
+    "      embedding_model: sentence-transformers/multi-qa-mpnet-base-dot-v1\n",
+    "```\n",
+    "\n",
+    "After defining components, create a query pipeline in `pipelines` section. Here, `name` refers to the name of the pipeline and `nodes` defines how the pipeline is built. \n",
````
```suggestion
    "After you define the nodes, create a query pipeline in the `pipelines` section. Here, `name` refers to the name of the pipeline, and `nodes` defines the order of the nodes in the pipeline: \n",
```

agnieszka-m commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,538 @@
+   "source": [
+    "2. Create an indexing pipeline.\n",
+    "\n",
+    "You can define an indexing pipeline in the same pipeline YAML file and index your documents to Elasticsearch through REST API. For that, create `FileTypeClassifier`, `TextConverter`, `PDFToTextConverter`, and last, `PreProcessor` nodes with parameters to split documents. \n",
```
```suggestion
    "You can define an indexing pipeline in the same pipeline YAML file and index your documents to Elasticsearch through REST API. For that, create `FileTypeClassifier`, `TextConverter`, `PDFToTextConverter`, and  `PreProcessor` nodes. For `PreProcessor`, use `params` to define how you want to split your documents: \n",
```

agnieszka-m commented on tutorials/20_REST_API.ipynb:
````diff
@@ -0,0 +1,538 @@
+    "      split_respect_sentence_boundary: True \n",
+    "```\n",
+    "\n",
+    "Then, go below to the `pipelines` section of the YAML file and create a new pipeline called `indexing`. In this pipeline, indicate how these nodes are connected to each other, Retriever, and DocumentStore. This indexing pipeline supports `.txt` and `.pdf` files and pre-processes them before loading to the Elasticsearch.\n",
````
```suggestion
    "Then, in the `pipelines` section of the YAML file, create a new pipeline called `indexing`. In this pipeline, indicate how the nodes you just defined are connected to each other, Retriever, and DocumentStore. This indexing pipeline supports `.TXT` and `.PDF` files and pre-processes them before loading to the Elasticsearch.\n",
```

Acc. to our guidelines, we write file formats in all caps.

agnieszka-m commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,538 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "After completing query and indexing pipelines, add `version: ignore` to the top of the file. Now, the pipeline YAML is completed.\n",
```
```suggestion
    "After completing query and indexing pipelines, add `version: ignore` to the top of the file. Now, the pipeline YAML is ready.\n",
```

agnieszka-m commented on tutorials/20_REST_API.ipynb:
````diff
@@ -0,0 +1,538 @@
+    "        inputs: [Retriever]\n",
+    "```\n",
+    "\n",
+    "Feel free to play with the pipeline setup later on. Add or remove some nodes, change the parameters or add new ones. Check out [Haystack API Reference](https://docs.haystack.deepset.ai/reference/answer-generator-api) for more options on nodes and parameters."
````
```suggestion
    "Feel free to play with the pipeline setup later on. Add or remove some nodes, change the parameters, or add new ones. For more options for nodes and parameters, check out [Haystack API Reference](https://docs.haystack.deepset.ai/reference/answer-generator-api)."
```

With links, we always specify the purpose first, so " To learn about blah, go to blah" :)

agnieszka-m commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,538 @@
+   "source": [
+    "## Start Haystack API\n",
+    "\n",
+    "Now, you can start the REST API server running the pipelines above with gunicorn server. When the startup is complete, you will see it running on port 8000."
```
```suggestion
    "Now, you can start the REST API server running the pipelines above with the gunicorn server. When the startup is complete, you will see it running on port 8000."
```

agnieszka-m commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,538 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Before continuing, test if everything is okay with the Haystack API by sending a cURL request to the `/initialized` endpoint. You can use command line of your computer or tools like [Postman](https://www.postman.com/) to send cURL requests. If there is no problem, you will get a response as `true`."
```
```suggestion
    "Before continuing, test if everything is OK with the Haystack API by sending a cURL request to the `/initialized` endpoint. You can use the command line of your computer or tools like [Postman](https://www.postman.com/) to send cURL requests. If everything works fine, you will get `true` as a response."
```

agnieszka-m commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,538 @@
+   "source": [
+    "2. Index files to Elasticsearch.\n",
+    "\n",
+    "You can send cURL requests to the `/file-upload` endpoint to upload files to the Elasticsearch instance. Replace `<PATH_TO_FOLDER>` with the path to the example files on your computer and send the POST request. If the file is successfully uploaded, you will get a response as `null`."
```
```suggestion
    "You can send cURL requests to the `/file-upload` endpoint to upload files to the Elasticsearch instance. Replace `<PATH_TO_FOLDER>` with the path to the example files on your computer and send the POST request. If the file is successfully uploaded, you will get `null` as a response."
```

agnieszka-m commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,538 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "However, this method is not convenient for uploading multiple files to the Elasticsearch instance as replacing the file names in the request by hand is difficult. Instead, create a python file in the folder that you keep the example files, put the code below into the python file, and run the python script. This python code takes the name of every file in the folder and sends the POST request ...
```
```suggestion
    "This method is not convenient for uploading multiple files to the Elasticsearch instance as replacing the file names in the request by hand is difficult. Instead, create a python file in the folder where you keep the example files, put the code below into the python file, and run the python script. This python code takes the name of every file in the folder and sends the POST request to the URL `http://127.0.0.1:8000/file-upload`. With the `print(response.text, \":\", file_name)` line, you will be able to see the names of all indexed files. Make sure you see the \"Completed\" text before continuing. \n",
```

agnieszka-m commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,538 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    " As response, you will get a `QueryResponse` object consists of `query`, `answers`, and `documents`. Related Documents to your query will be under the `documents` attribute of the object.\n",
```
```suggestion
    " As a response, you will get a `QueryResponse` object consisting of `query`, `answers`, and `documents`. Documents related to your query will be under the `documents` attribute of the object.\n",
```

agnieszka-m reviewed (commented):
Just some language comments. Great job!

masci commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,538 @@
+    "## Overview\n",
+    "\n",
+    "Haystack enables you to apply the latest NLP technology to your own data and create production-ready applications. Building an end-to-end NLP application requires the combination of multiple concepts:\n",
+    "* **DocumentStore** stores the data. You will use Elasticsearch for this tutorial.\n",
```
If you are not familiar with Haystack this doesn't tell you much: I'd say something along the lines of 
> DocumentStore: is the component in Haystack responsible for loading and storing text data in form of Documents. In this tutorial, the DocumentStore will use Elasticsearch behind the scene

masci commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,538 @@
+    "\n",
+    "Haystack enables you to apply the latest NLP technology to your own data and create production-ready applications. Building an end-to-end NLP application requires the combination of multiple concepts:\n",
+    "* **DocumentStore** stores the data. You will use Elasticsearch for this tutorial.\n",
+    "* **Haystack** pipelines convert files into [Documents](https://docs.haystack.deepset.ai/docs/documents_answers_labels#document), index them to DocumentStore, and run NLP tasks such as question answering and document search.\n",
```
Same, maybe something like
> Pipelines are a set of Haystack components connected together to achieve a specific goal, like convert files into Documents and store them, or run NLP tasks such as...

masci commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,538 @@
+    "Haystack enables you to apply the latest NLP technology to your own data and create production-ready applications. Building an end-to-end NLP application requires the combination of multiple concepts:\n",
+    "* **DocumentStore** stores the data. You will use Elasticsearch for this tutorial.\n",
+    "* **Haystack** pipelines convert files into [Documents](https://docs.haystack.deepset.ai/docs/documents_answers_labels#document), index them to DocumentStore, and run NLP tasks such as question answering and document search.\n",
+    "* **REST API** allows other applications to interact with Haystack by handling their queries and returning responses.\n",
```
If we refer to the Python app/package, we should use `rest_api`:
> `rest_api` is a tiny application that exposes Haystack's functionalities through a RESTful API, for example accepting queries and returning results

masci commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,538 @@
+   "source": [
+    "%%bash\n",
+    "\n",
+    "apt-get update && apt-get install docker && apt-get install docker-compose\n",
```
I understand this tutorial is not meant to be run on Colab - in this case, I would not provide `bash` cells. For example, this one would fail on the average Mac setup, while another one below would fail on Linux. We should probably just point users towards the Docker install page. Something we could do, is adding a cell users can run to confirm Docker is up and running, like `docker info`

masci commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,538 @@
+   "source": [
+    "2. Install Haystack.\n",
+    "\n",
+    "Install the latest version of Haystack from the main branch and all its dependencies. REST API and `xpdf` are also required.  "
```
`rest_api`

masci commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,538 @@
+   "source": [
+    "4. Update the `docker-compose.yml` file.\n",
+    "\n",
+    "The current `docker-compose.yml` file has basic configs and it also provides an image for Elasticsearch container which needs to change. For that, go to `docker-compose.yml` and replace `image: \"deepset/elasticsearch-countries-and-capitals\"` with `image: \"docker.elastic.co/elasticsearch/elasticsearch:7.9.2\"`. The new image is going to provide an empty Elasticsearch instance instead of an ...
```
This feels a bit confusing, modifying a compose file that's meant for another app. 

Also if we only run one container (ES) there's no need for a compose file at all, you could just `docker run` Elasticsearch. So if we want to use a compose file, I would also use the Haystack image already shipping `rest_api` and `xpdf`, that would simplify the overall setup. Also I would use a dedicated compose file users can copypaste without modification.

And if we go

masci commented on tutorials/20_REST_API.ipynb:
```diff
@@ -0,0 +1,538 @@
+    "services:\n",
+    "  ...\n",
+    "  elasticsearch:\n",
+    "    image: \"docker.elastic.co/elasticsearch/elasticsearch:7.9.2\"\n",
```
let's use a more recent 7.x, like `elasticsearch:7.17.6`

masci reviewed (changes requested):
I love this tutorial, this is very much needed!

I took a first pass and I stopped at the setup phase, let me know what you think!

masci commented on tutorials/20_Using_Haystack_with_REST_API.ipynb:
````diff
@@ -0,0 +1,415 @@
+    "\n",
+    "```python\n",
+    "import os\n",
+    "import requests\n",
````
this is going to be tricky: `requests` needs to be installed but I'm not sure we want to go through the rabbit hole of setting up a Python virtual env. I suggest we either use a "pure python" solution or a shell command, for example (not tested):
```
find ./files -name '*.txt' -exec curl -s -X POST -H 'Accept: application/json' -F files="@{}" http://127.0.0.1:8000/file-upload;
```

bilgeyucel commented on tutorials/20_Using_Haystack_with_REST_API.ipynb:
```diff
@@ -0,0 +1,415 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "2. Clone Haystack repository.\n",
```
Hi @masci, I agree that they don't need to clone the whole repository only for the compose file, yet, I am not sure how to mitigate this issue. In my opinion, if we make them create the compose file from scratch in the tutorial, we need to explain what each line of code does. I didn't want to do that because:
1. Explaining the compose file would make the tutorial longer and more complicated for inexperienced users
2. The focus point of this tutorial is the pipeline YAML file and how to edit it, I don't want to shift toward Docker and put it in a spotlight

That's why I think for this tutorial it's more important to provide a compose file that works out of the box. Maybe I can give the Github link to the compose file and ask them to copy/paste it instead of making them clone the repo. wdyt?

masci replied:
Good point, they can `curl https://raw.githubusercontent.com/deepset-ai/haystack/main/docker-compose.yml` 👍

bilgeyucel commented on tutorials/20_Using_Haystack_with_REST_API.ipynb:
```diff
@@ -0,0 +1,415 @@
+   "source": [
+    "1. Create a document search pipeline.\n",
+    "\n",
+    "Time to design a document search pipeline from scratch. This will be your query pipeline. Create a new file named `document-search.haystack-pipeline.yml` in the `/pipeline` folder under `/rest_api` in the Haystack code base. Then, create a `PIPELINE_YAML_PATH` variable in the `docker-compose.yml` with the new file name. `PIPELINE_YAML_PATH` variable will tell `rest_api` which YAML file to ru ...
```
This is a great idea 💯

TuanaCelik commented on markdowns/20_Using_Haystack_with_REST_API.md:
````diff
@@ -0,0 +1,312 @@
+   ```
+
+2. Download the *docker-compose.yml* file. Haystack provides a *docker-compose.yml* file that defines services for Haystack API and Elasticsearch. 
+    1. Create a new folder called *doc-search*.
````
Cosmetic, not too important so take or leave this comment: 
As a developer my first question might be 'where'? So maybe somewhere above I would mention that these are the steps for a brand new project we have set up.

TuanaCelik commented on markdowns/20_Using_Haystack_with_REST_API.md:
````diff
@@ -0,0 +1,312 @@
+              inputs: [Retriever]
+      ```
+
+3. After creating query and indexing pipelines, add `version: ignore` to the top of the file. Now, the pipeline YAML is ready.
````
Again, maybe too much detail and nitpicking but in paranthesis maybe a note on why we `ignore`?

TuanaCelik commented on markdowns/20_Using_Haystack_with_REST_API.md:
```diff
@@ -0,0 +1,312 @@
+- [FARM](https://github.com/deepset-ai/FARM)
+
+Get in touch:
+[Twitter](https://twitter.com/deepset_ai) | [LinkedIn](https://www.linkedin.com/company/deepset-ai/) | [Discord](https://haystack.deepset.ai/community/join) | [GitHub Discussions](https://github.com/deepset-ai/haystack/discussions) | [Website](https://deepset.ai)
```
Change https://haystack.deepset.ai/community/join to https://haystack.deepset.ai/community

masci reviewed (approved):
LGTM! 👍

anakin87 reviewed (approved):
Looks very good to me!
